fvm/meter/interaction_meter.go

```diff
@@ -51,34 +59,68 @@ func (m *InteractionMeter) MeterStorageRead(
 ) error {
 
 	// all reads are on a View which only read from storage at the first read of a given key
-	if _, ok := m.storageUpdateSizeMap[storageKey]; !ok {
+	if _, ok := m.reads[storageKey]; !ok {
 		readByteSize := getStorageKeyValueSize(storageKey, value)
 		m.totalStorageBytesRead += readByteSize
-		m.storageUpdateSizeMap[storageKey] = readByteSize
+		m.reads[storageKey] = readByteSize
 	}
 
 	return m.checkStorageInteractionLimit(enforceLimit)
 }
 
 // MeterStorageWrite captures storage written bytes count and returns an error
-// if it goes beyond the total interaction limit and limit is enforced
+// if it goes beyond the total interaction limit and limit is enforced.
+// If a key is written multiple times, only the last write is counted.
+// If a key is written before it has been read, next time it will be read it will be from the view,
+// not from storage, so count it as read 0.
 func (m *InteractionMeter) MeterStorageWrite(
 	storageKey flow.RegisterID,
 	value flow.RegisterValue,
 	enforceLimit bool,
 ) error {
-	// all writes are on a View which only writes the latest updated value to storage at commit
-	if old, ok := m.storageUpdateSizeMap[storageKey]; ok {
-		m.totalStorageBytesWritten -= old
-	}
-
 	updateSize := getStorageKeyValueSize(storageKey, value)
-	m.totalStorageBytesWritten += updateSize
-	m.storageUpdateSizeMap[storageKey] = updateSize
+	m.replaceWrite(storageKey, updateSize)
+
+	if _, ok := m.reads[storageKey]; !ok {
+		// write without read, count as read 0 because next time you read it the written value will be returned
+		// so no interaction with storage will happen.
+		m.reads[storageKey] = 0
+	}
 
 	return m.checkStorageInteractionLimit(enforceLimit)
 }
 
+// replaceWrite replaces the write size of a given key with the new size, because
+// only the last write of a given key is counted towards the total interaction limit.
+// These are the only write usages of `m.totalStorageBytesWritten` and `m.writes`,
+// which means that `m.totalStorageBytesWritten` can never become negative.
+// oldSize is always <= m.totalStorageBytesWritten.
+func (m *InteractionMeter) replaceWrite(
+	k flow.RegisterID,
+	newSize uint64,
+) {
+	totalBefore := m.totalStorageBytesWritten
+
+	// remove old write
+	oldSize := m.writes[k]
+	m.totalStorageBytesWritten -= oldSize
+
+	// sanity check
+	// this should never happen, but if it does, we want to know about it
+	if m.totalStorageBytesWritten > totalBefore {
+		log.Error().
+			Str("component", "interaction_meter").
+			Uint64("total", totalBefore).
+			Uint64("subtract", oldSize).
+			Msg("totalStorageBytesWritten would have become negative")
+		m.totalStorageBytesWritten = 0
```

Comment:
Is this a fatal error? if so, we should just crash the node, rather than log and continue. This type of error is easy to be missed and ignored. 
```suggestion
		log.Fatal().
			Str("component", "interaction_meter").
			Uint64("total", totalBefore).
			Uint64("subtract", oldSize).
			Msg("totalStorageBytesWritten would have become negative")
```

Reply:
I would not treat it as fatal, as it only influences how quickly transactions hit the interaction limit, and nothing else. On the other hand, this should not (is not able to) happen so it could be fatal.

I'm a bit torn about it, but I decided to go with error. What do you think?

Reply:
I still think this should be a `Fatal` error. As the reason running into here is unknown, in the worst case, some node ran into this error, and some didn't, then different nodes would produce different transaction results, which create a result fork. That would take much bigger effort to fix than simply crashing the node here, where all nodes are stopped at the same height with a consistent state.